File: packages/server/src/domains/users/repo.ts
import { getOne, getMany } from '../../shared/db.js';
import type { User, CreateUserInput, UpdateUserInput, CreateSamlUserInput } from '@flick/shared';

interface UserRow extends User {
  password_hash: string | null;
  saml_name_id: string | null;
  saml_issuer: string | null;
}

const USER_COLUMNS = 'id, email, name, role, auth_method, created_at, updated_at';

export async function findAll(): Promise<User[]> {
  return getMany<User>(`SELECT ${USER_COLUMNS} FROM users ORDER BY created_at DESC`);
}

export async function findById(id: string): Promise<User | null> {
  return getOne<User>(`SELECT ${USER_COLUMNS} FROM users WHERE id = $1`, [id]);
}

export async function findByIdWithHash(id: string): Promise<UserRow | null> {
  return getOne<UserRow>('SELECT * FROM users WHERE id = $1', [id]);
}

export async function findByEmail(email: string): Promise<UserRow | null> {
  return getOne<UserRow>('SELECT * FROM users WHERE email = $1', [email]);
}

export async function findBySamlIdentity(nameId: string, issuer: string): Promise<UserRow | null> {
  return getOne<UserRow>(
    'SELECT * FROM users WHERE saml_name_id = $1 AND saml_issuer = $2',
    [nameId, issuer],
  );
}

export async function create(input: CreateUserInput & { password_hash: string }): Promise<User> {
  const result = await getOne<User>(
    `INSERT INTO users (email, name, password_hash, role, auth_method)
     VALUES ($1, $2, $3, $4, 'password')
     RETURNING ${USER_COLUMNS}`,
    [input.email, input.name, input.password_hash, input.role],
  );
  return result!;
}

export async function createSamlUser(input: CreateSamlUserInput): Promise<User> {
  const result = await getOne<User>(
    `INSERT INTO users (email, name, auth_method, saml_name_id, saml_issuer, role)
     VALUES ($1, $2, 'saml', $3, $4, 'viewer')
     RETURNING ${USER_COLUMNS}`,
    [input.email, input.name, input.saml_name_id, input.saml_issuer],
  );
  return result!;
}

export async function update(id: string, input: UpdateUserInput): Promise<User | null> {
  const fields: string[] = [];
  const values: unknown[] = [];
  let idx = 1;

  if (input.name !== undefined) { fields.push(`name = $${idx++}`); values.push(input.name); }
  if (input.email !== undefined) { fields.push(`email = $${idx++}`); values.push(input.email); }
  if (input.role !== undefined) { fields.push(`role = $${idx++}`); values.push(input.role); }

  if (fields.length === 0) return findById(id);

  values.push(id);
  return getOne<User>(
    `UPDATE users SET ${fields.join(', ')} WHERE id = $${idx}
     RETURNING ${USER_COLUMNS}`,
    values,
  );
}

export async function updateProfile(id: string, input: { name?: string; email?: string }): Promise<User | null> {
  const fields: string[] = [];
  const values: unknown[] = [];
  let idx = 1;

  if (input.name !== undefined) { fields.push(`name = $${idx++}`); values.push(input.name); }
  if (input.email !== undefined) { fields.push(`email = $${idx++}`); values.push(input.email); }

  if (fields.length === 0) return findById(id);

  values.push(id);
  return getOne<User>(
    `UPDATE users SET ${fields.join(', ')} WHERE id = $${idx}
     RETURNING ${USER_COLUMNS}`,
    values,
  );
}

export async function updatePasswordHash(id: string, passwordHash: string): Promise<boolean> {
  const result = await getOne<{ id: string }>(
    'UPDATE users SET password_hash = $1 WHERE id = $2 RETURNING id',
    [passwordHash, id],
  );
  return result !== null;
}

export async function remove(id: string): Promise<boolean> {
  const result = await getOne<{ id: string }>('DELETE FROM users WHERE id = $1 RETURNING id', [id]);
  return result !== null;
}

File: packages/server-rust/src/domains/users/repo.rs
use chrono::{DateTime, Utc};
use serde::Serialize;
use sqlx::{FromRow, PgPool};
use uuid::Uuid;

const USER_COLUMNS: &str = "id, email, name, role, auth_method, created_at, updated_at";

// ── Row types ──────────────────────────────────────────────────────────────

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct UserRow {
    pub id: Uuid,
    pub email: String,
    pub name: String,
    pub role: String,
    pub auth_method: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow)]
pub struct UserWithHashRow {
    pub id: Uuid,
    pub email: String,
    pub name: String,
    pub role: String,
    pub auth_method: String,
    pub password_hash: Option<String>,
    pub saml_name_id: Option<String>,
    pub saml_issuer: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

// ── Queries ────────────────────────────────────────────────────────────────

pub async fn find_all(db: &PgPool) -> Result<Vec<UserRow>, sqlx::Error> {
    let sql = format!(
        "SELECT {} FROM users ORDER BY created_at DESC",
        USER_COLUMNS
    );
    sqlx::query_as::<_, UserRow>(&sql).fetch_all(db).await
}

pub async fn find_by_id(db: &PgPool, id: Uuid) -> Result<Option<UserRow>, sqlx::Error> {
    let sql = format!("SELECT {} FROM users WHERE id = $1", USER_COLUMNS);
    sqlx::query_as::<_, UserRow>(&sql)
        .bind(id)
        .fetch_optional(db)
        .await
}

pub async fn find_by_id_with_hash(
    db: &PgPool,
    id: Uuid,
) -> Result<Option<UserWithHashRow>, sqlx::Error> {
    sqlx::query_as::<_, UserWithHashRow>("SELECT * FROM users WHERE id = $1")
        .bind(id)
        .fetch_optional(db)
        .await
}

pub async fn find_by_email(
    db: &PgPool,
    email: &str,
) -> Result<Option<UserWithHashRow>, sqlx::Error> {
    sqlx::query_as::<_, UserWithHashRow>("SELECT * FROM users WHERE email = $1")
        .bind(email)
        .fetch_optional(db)
        .await
}

pub async fn find_by_saml_identity(
    db: &PgPool,
    name_id: &str,
    issuer: &str,
) -> Result<Option<UserWithHashRow>, sqlx::Error> {
    sqlx::query_as::<_, UserWithHashRow>(
        "SELECT * FROM users WHERE saml_name_id = $1 AND saml_issuer = $2",
    )
    .bind(name_id)
    .bind(issuer)
    .fetch_optional(db)
    .await
}

pub async fn create(
    db: &PgPool,
    email: &str,
    name: &str,
    password_hash: &str,
    role: &str,
) -> Result<UserRow, sqlx::Error> {
    let sql = format!(
        r#"
        INSERT INTO users (id, email, name, password_hash, role, auth_method, created_at, updated_at)
        VALUES (gen_random_uuid(), $1, $2, $3, $4, 'password', now(), now())
        RETURNING {}
        "#,
        USER_COLUMNS
    );
    sqlx::query_as::<_, UserRow>(&sql)
        .bind(email)
        .bind(name)
        .bind(password_hash)
        .bind(role)
        .fetch_one(db)
        .await
}

pub async fn create_saml_user(
    db: &PgPool,
    email: &str,
    name: &str,
    saml_name_id: &str,
    saml_issuer: &str,
) -> Result<UserRow, sqlx::Error> {
    let sql = format!(
        r#"
        INSERT INTO users (id, email, name, role, auth_method, saml_name_id, saml_issuer, created_at, updated_at)
        VALUES (gen_random_uuid(), $1, $2, 'viewer', 'saml', $3, $4, now(), now())
        RETURNING {}
        "#,
        USER_COLUMNS
    );
    sqlx::query_as::<_, UserRow>(&sql)
        .bind(email)
        .bind(name)
        .bind(saml_name_id)
        .bind(saml_issuer)
        .fetch_one(db)
        .await
}

pub async fn update(
    db: &PgPool,
    id: Uuid,
    name: Option<&str>,
    email: Option<&str>,
    role: Option<&str>,
) -> Result<Option<UserRow>, sqlx::Error> {
    let mut sets: Vec<String> = Vec::new();
    let mut param_index: u32 = 1;

    if name.is_some() {
        sets.push(format!("name = ${param_index}"));
        param_index += 1;
    }
    if email.is_some() {
        sets.push(format!("email = ${param_index}"));
        param_index += 1;
    }
    if role.is_some() {
        sets.push(format!("role = ${param_index}"));
        param_index += 1;
    }

    if sets.is_empty() {
        return find_by_id(db, id).await;
    }

    sets.push("updated_at = now()".to_string());

    let sql = format!(
        "UPDATE users SET {} WHERE id = ${} RETURNING {}",
        sets.join(", "),
        param_index,
        USER_COLUMNS
    );

    let mut query = sqlx::query_as::<_, UserRow>(&sql);

    if let Some(v) = name {
        query = query.bind(v);
    }
    if let Some(v) = email {
        query = query.bind(v);
    }
    if let Some(v) = role {
        query = query.bind(v);
    }
    query = query.bind(id);

    query.fetch_optional(db).await
}

pub async fn update_profile(
    db: &PgPool,
    id: Uuid,
    name: Option<&str>,
    email: Option<&str>,
) -> Result<Option<UserRow>, sqlx::Error> {
    let mut sets: Vec<String> = Vec::new();
    let mut param_index: u32 = 1;

    if name.is_some() {
        sets.push(format!("name = ${param_index}"));
        param_index += 1;
    }
    if email.is_some() {
        sets.push(format!("email = ${param_index}"));
        param_index += 1;
    }

    if sets.is_empty() {
        return find_by_id(db, id).await;
    }

    sets.push("updated_at = now()".to_string());

    let sql = format!(
        "UPDATE users SET {} WHERE id = ${} RETURNING {}",
        sets.join(", "),
        param_index,
        USER_COLUMNS
    );

    let mut query = sqlx::query_as::<_, UserRow>(&sql);

    if let Some(v) = name {
        query = query.bind(v);
    }
    if let Some(v) = email {
        query = query.bind(v);
    }
    query = query.bind(id);

    query.fetch_optional(db).await
}

pub async fn update_password_hash(
    db: &PgPool,
    id: Uuid,
    password_hash: &str,
) -> Result<(), sqlx::Error> {
    sqlx::query("UPDATE users SET password_hash = $1, updated_at = now() WHERE id = $2")
        .bind(password_hash)
        .bind(id)
        .execute(db)
        .await?;
    Ok(())
}

pub async fn remove(db: &PgPool, id: Uuid) -> Result<bool, sqlx::Error> {
    let result = sqlx::query("DELETE FROM users WHERE id = $1")
        .bind(id)
        .execute(db)
        .await?;
    Ok(result.rows_affected() > 0)
}
